internal/db/db_certificate_authorities.go

```diff
@@ -274,3 +336,99 @@ func (db *Database) DeleteCertificateAuthority(filter CertificateAuthorityFilter
+// The CSR filter finds the CSR to sign. the CA Filter finds the CA that will issue the certificate.
+func (db *Database) SignCertificateRequest(csrFilter CSRFilter, caFilter CertificateAuthorityFilter) error {
+	csrRow, err := db.GetCertificateRequest(csrFilter)
+	if err != nil {
+		return err
+	}
+	caRow, err := db.GetDenormalizedCertificateAuthority(caFilter)
+	if err != nil {
+		return err
+	}
+	if csrRow.CertificateID != 0 {
+		return errors.New("CSR already signed. please renew instead")
+	}
+	if caRow.CertificateChain == "" {
+		return errors.New("CA does not have a valid signed certificate to sign certificates")
+	}
+	if caRow.Status != CAActive {
+		return errors.New("CA is not active to sign certificates")
+	}
+	block, _ := pem.Decode([]byte(csrRow.CSR))
+	certRequest, err := x509.ParseCertificateRequest(block.Bytes)
+	if err != nil {
+		return err
+	}
+	certChain, err := sanitizeCertificateBundle(caRow.CertificateChain)
+	if err != nil {
+		return err
+	}
+	block, _ = pem.Decode([]byte(certChain[0]))
+	caCertParsed, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return err
+	}
+	block, _ = pem.Decode([]byte(caRow.PrivateKeyPEM))
+	caPrivateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return err
+	}
+	if err = certRequest.CheckSignature(); err != nil {
+		return fmt.Errorf("invalid certificate request signature: %w", err)
+	}
+	CSRIsForACertificateAuthority := true
+	caToBeSigned, err := db.GetCertificateAuthority(ByCertificateAuthorityCSRID(csrRow.CSR_ID))
+	if err != nil {
+		if !errors.Is(err, sqlair.ErrNoRows) {
+			return err
+		}
+		CSRIsForACertificateAuthority = false
+	}
+	// Create certificate template from the CSR
+	certTemplate := &x509.Certificate{
+		Subject:            certRequest.Subject,
+		EmailAddresses:     certRequest.EmailAddresses,
+		IPAddresses:        certRequest.IPAddresses,
+		URIs:               certRequest.URIs,
+		DNSNames:           certRequest.DNSNames,
+		PublicKey:          certRequest.PublicKey,
+		PublicKeyAlgorithm: certRequest.PublicKeyAlgorithm,
+
+		// Add standard certificate fields
+		SerialNumber: big.NewInt(time.Now().UnixNano()),
+		NotBefore:    time.Now(),
+		NotAfter:     time.Now().AddDate(1, 0, 0),
+		KeyUsage:     x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+	}
+	if CSRIsForACertificateAuthority {
+		certTemplate.KeyUsage = x509.KeyUsageCertSign | x509.KeyUsageCRLSign
+		certTemplate.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
+		certTemplate.BasicConstraintsValid = true
+		certTemplate.IsCA = true
+	}
+	certBytes, err := x509.CreateCertificate(rand.Reader, certTemplate, caCertParsed, certTemplate.PublicKey, caPrivateKey)
+	if err != nil {
+		return err
+	}
+	certPEM := new(bytes.Buffer)
+	err = pem.Encode(certPEM, &pem.Block{Type: "CERTIFICATE", Bytes: certBytes})
```

Comment:
Ain't there an enum that we can use (from the `pem` library) to specify we want a certificate instead of using a string?